File: 0c-bason/assignment-1/cpp/src/ron64.cpp
#include "ron64.h"

#include <stdexcept>
#include <algorithm>

////////////////////////////////////////////////////////////////////////////////

namespace NBason {

////////////////////////////////////////////////////////////////////////////////

namespace {

// RON64 alphabet in ASCII order: 0-9A-Z_a-z~
constexpr char RON64_ALPHABET[] = 
    "0123456789"                  // 0-9
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"  // 10-35
    "_"                           // 36
    "abcdefghijklmnopqrstuvwxyz"  // 37-62
    "~";                          // 63

constexpr size_t RON64_BASE = 64;

// Build reverse lookup table for decoding
constexpr int BuildDecodeTable(char ch) {
    if (ch >= '0' && ch <= '9') return ch - '0';
    if (ch >= 'A' && ch <= 'Z') return 10 + (ch - 'A');
    if (ch == '_') return 36;
    if (ch >= 'a' && ch <= 'z') return 37 + (ch - 'a');
    if (ch == '~') return 63;
    return -1;
}

} // namespace

////////////////////////////////////////////////////////////////////////////////

std::string EncodeRon64(uint64_t value)
{
    // Special case: zero
    if (value == 0) {
        return "0";
    }

    // Convert to base-64 (big-endian: most significant digit first)
    std::string result;
    while (value > 0) {
        result += RON64_ALPHABET[value % RON64_BASE];
        value /= RON64_BASE;
    }

    // Reverse to get big-endian order
    std::reverse(result.begin(), result.end());
    
    return result;
}

uint64_t DecodeRon64(const std::string& str)
{
    if (str.empty()) {
        throw std::invalid_argument("Empty RON64 string");
    }

    uint64_t result = 0;
    
    for (char ch : str) {
        int digit = BuildDecodeTable(ch);
        if (digit < 0) {
            throw std::invalid_argument("Invalid RON64 character: " + std::string(1, ch));
        }

        // Check for overflow
        if (result > (UINT64_MAX / RON64_BASE)) {
            throw std::overflow_error("RON64 value overflow");
        }
        
        result = result * RON64_BASE + digit;
    }

    return result;
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NBason

File: 0c-bason/assignment-1/go/pkg/bason/ron64.go
package bason

import (
	"errors"
	"fmt"
)

// RON64 alphabet in ASCII order: 0-9A-Z_a-z~
const ron64Alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz~"
const ron64Base = 64

var (
	ErrEmptyRon64   = errors.New("empty RON64 string")
	ErrInvalidRon64 = errors.New("invalid RON64 character")
	ErrRon64Overflow = errors.New("RON64 value overflow")
)

func buildDecodeTable(ch byte) int {
	if ch >= '0' && ch <= '9' {
		return int(ch - '0')
	}
	if ch >= 'A' && ch <= 'Z' {
		return 10 + int(ch-'A')
	}
	if ch == '_' {
		return 36
	}
	if ch >= 'a' && ch <= 'z' {
		return 37 + int(ch-'a')
	}
	if ch == '~' {
		return 63
	}
	return -1
}

// EncodeRon64 encodes a non-negative integer to a RON64 string.
// Numbers are encoded in big-endian digit order (most significant first).
func EncodeRon64(value uint64) string {
	if value == 0 {
		return "0"
	}
	var result []byte
	for value > 0 {
		result = append(result, ron64Alphabet[value%ron64Base])
		value /= ron64Base
	}
	// Reverse to get big-endian order
	for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = result[j], result[i]
	}
	return string(result)
}

// DecodeRon64 decodes a RON64 string to an integer.
// Returns an error on invalid characters or overflow.
func DecodeRon64(str string) (uint64, error) {
	if len(str) == 0 {
		return 0, ErrEmptyRon64
	}
	var result uint64
	for i := 0; i < len(str); i++ {
		digit := buildDecodeTable(str[i])
		if digit < 0 {
			return 0, fmt.Errorf("%w: %q", ErrInvalidRon64, str[i])
		}
		const maxU64 = ^uint64(0)
		if result > maxU64/ron64Base {
			return 0, ErrRon64Overflow
		}
		result = result*ron64Base + uint64(digit)
	}
	return result, nil
}
